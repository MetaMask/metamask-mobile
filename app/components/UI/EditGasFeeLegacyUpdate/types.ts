import { GasFeeOptions } from '../../../core/GasPolling/types';

export interface EditGasFeeLegacyUpdateProps {
  /**
   * Gas option selected (low, medium, high)
   */
  selected: string;
  /**
   * Gas fee options to select from
   */
  gasOptions: GasFeeOptions;
  /**
   * Function called when user selected or changed the gas
   */
  onChange: any;
  /**
   * Function called when user cancels
   */
  onCancel: any;
  /**
   * Function called when user saves the new gas
   */
  onSave: any;
  /**
   * Primary currency, either ETH or Fiat
   */
  primaryCurrency: string;
  /**
   * A string representing the network chainId
   */
  chainId: string;
  /**
   * Estimate type returned by the gas fee controller, can be market-fee, legacy or eth_gasPrice
   */
  gasEstimateType: string;

  /**
   * Error message to show
   */
  error: any;
  /**
   * Warning message to show
   */
  warning: any | undefined;
  /**
   * Ignore option array
   */
  ignoreOptions: string[] | undefined;
  /**
   * Extend options object. Object has option keys and properties will be spread
   */
  extendOptions: any;
  /**
   * Recommended object with type and render function
   */
  recommended: any;
  /**
   * Estimate option to compare with for too low warning
   */
  warningMinimumEstimateOption: string;
  /**
   * Function to call when update animation starts
   */
  onUpdatingValuesStart: () => void;
  /**
   * Function to call when update animation ends
   */
  onUpdatingValuesEnd: () => void;
  /**
   * If the values should animate upon update or not
   */
  animateOnChange: boolean | undefined;
  /**
   * Boolean to determine if the animation is happening
   */
  isAnimating: boolean;
  /**
   * Extra analytics params to be send with the gas analytics
   */
  analyticsParams: any;
  view: string;
  onlyGas?: boolean;
  selectedGasObject: any;
}
