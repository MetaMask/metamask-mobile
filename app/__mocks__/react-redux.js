export * from 'react-redux-test';
