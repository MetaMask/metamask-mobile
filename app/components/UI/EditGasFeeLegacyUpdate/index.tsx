/* eslint-disable react/display-name */
import React, { useCallback, useState, useMemo } from 'react';
import {
  View,
  TouchableOpacity,
  ScrollView,
  TouchableWithoutFeedback,
} from 'react-native';
import { GAS_ESTIMATE_TYPES } from '@metamask/gas-fee-controller';
import BigNumber from 'bignumber.js';
import Text from '../../Base/Text';
import StyledButton from '../StyledButton';
import RangeInput from '../../Base/RangeInput';
import MaterialCommunityIcon from 'react-native-vector-icons/MaterialCommunityIcons';
import InfoModal from '../Swaps/components/InfoModal';
import Icon from 'react-native-vector-icons/Ionicons';
import { strings } from '../../../../locales/i18n';
import Alert, { AlertType } from '../../Base/Alert';
import HorizontalSelector from '../../Base/HorizontalSelector';
import { isMainnetByChainId } from '../../../util/networks';
import FadeAnimationView from '../FadeAnimationView';
import { MetaMetricsEvents } from '../../../core/Analytics';
import AnalyticsV2 from '../../../util/analyticsV2';

import AppConstants from '../../../core/AppConstants';
import { useTheme } from '../../../util/theme';
import { EditGasFeeLegacyUpdateProps } from './types';
import { useGasTransaction } from '../../../core/GasPolling/GasPolling';
import createStyles from './styles';

const GAS_LIMIT_INCREMENT = new BigNumber(1000);
const GAS_PRICE_INCREMENT = new BigNumber(1);
const GAS_LIMIT_MIN = new BigNumber(21000);
const GAS_PRICE_MIN = new BigNumber(0);

const EditGasFeeLegacy = ({
  selected,
  gasOptions,
  onChange,
  onCancel,
  onSave,
  primaryCurrency,
  chainId,
  gasEstimateType,
  error,
  warning,
  ignoreOptions,
  extendOptions = {},
  recommended,
  warningMinimumEstimateOption,
  onUpdatingValuesStart,
  onUpdatingValuesEnd,
  animateOnChange,
  isAnimating,
  analyticsParams,
  view,
  onlyGas,
  selectedGasObject,
}: EditGasFeeLegacyUpdateProps) => {
  const onlyAdvanced = gasEstimateType !== GAS_ESTIMATE_TYPES.LEGACY;
  const [showRangeInfoModal, setShowRangeInfoModal] = useState<boolean>(false);
  const [showAdvancedOptions, setShowAdvancedOptions] = useState<boolean>(
    !selected || onlyAdvanced,
  );
  const [selectedOption, setSelectedOption] = useState<string>(selected);
  const [gasPriceError, setGasPriceError] = useState<string>('');
  const [gasObject, updateLegacyGasObject] = useState<{
    legacyGasLimit: string;
    suggestedGasPrice: string;
  }>({
    legacyGasLimit: selectedGasObject.legacyGasLimit,
    suggestedGasPrice: selectedGasObject.suggestedGasPrice,
  });
  const { colors } = useTheme();
  const styles = createStyles(colors);

  const gasTransaction = useGasTransaction({
    onlyGas,
    gasSelected: selectedOption,
    legacy: true,
    gasObject,
  });

  const getAnalyticsParams = useCallback(() => {
    try {
      return {
        ...analyticsParams,
        chain_id: chainId,
        function_type: view,
        gas_mode: selectedOption ? 'Basic' : 'Advanced',
        speed_set: selectedOption || undefined,
      };
    } catch (err) {
      return {};
    }
  }, [analyticsParams, chainId, selectedOption, view]);

  const toggleAdvancedOptions = useCallback(() => {
    if (!showAdvancedOptions) {
      AnalyticsV2.trackEvent(
        MetaMetricsEvents.GAS_ADVANCED_OPTIONS_CLICKED,
        getAnalyticsParams(),
      );
    }
    setShowAdvancedOptions(!showAdvancedOptions);
  }, [getAnalyticsParams, showAdvancedOptions]);

  const save = useCallback(() => {
    AnalyticsV2.trackEvent(
      MetaMetricsEvents.GAS_FEE_CHANGED,
      getAnalyticsParams(),
    );

    const newGasPriceObject = {
      suggestedGasPrice: gasObject?.suggestedGasPrice,
      legacyGasLimit: gasObject?.legacyGasLimit,
    };

    onSave(gasTransaction, newGasPriceObject, selectedOption);
  }, [getAnalyticsParams, onSave, gasTransaction, gasObject, selectedOption]);

  const changeGas = useCallback(
    (gas, option) => {
      setSelectedOption(option);
      updateLegacyGasObject({
        legacyGasLimit: gas.suggestedGasLimit,
        suggestedGasPrice: gas.suggestedGasPrice,
      });
      onChange(option);
    },
    [onChange],
  );

  const changedGasPrice = useCallback(
    (value: string) => {
      const lowerValue = new BigNumber(
        gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY
          ? gasOptions?.[warningMinimumEstimateOption]
          : gasOptions?.gasPrice,
      );
      const higherValue = new BigNumber(
        gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY
          ? gasOptions?.high
          : gasOptions?.gasPrice,
      ).multipliedBy(new BigNumber(1.5));

      const valueBN = new BigNumber(value);

      if (!lowerValue.isNaN() && valueBN.lt(lowerValue)) {
        setGasPriceError(strings('edit_gas_fee_eip1559.gas_price_low'));
      } else if (!higherValue.isNaN() && valueBN.gt(higherValue)) {
        setGasPriceError(strings('edit_gas_fee_eip1559.gas_price_high'));
      } else {
        setGasPriceError('');
      }

      const newGas = { ...gasTransaction, suggestedGasPrice: value };

      changeGas(newGas, null);
    },
    [
      changeGas,
      gasEstimateType,
      gasTransaction,
      gasOptions,
      warningMinimumEstimateOption,
    ],
  );

  const changedGasLimit = useCallback(
    (value: string) => {
      const newGas = { ...gasTransaction, suggestedGasLimit: value };
      changeGas(newGas, null);
    },
    [changeGas, gasTransaction],
  );

  const selectOption = useCallback(
    (option: string) => {
      setGasPriceError('');
      changeGas({ suggestedGasPrice: gasOptions[option] }, option);
    },
    [changeGas, gasOptions],
  );

  const shouldIgnore = useCallback(
    (option) => ignoreOptions.find((item) => item === option),
    [ignoreOptions],
  );

  const renderOptions = useMemo(
    () =>
      [
        {
          name: AppConstants.GAS_OPTIONS.LOW,
          label: strings('edit_gas_fee_eip1559.low'),
        },
        {
          name: AppConstants.GAS_OPTIONS.MEDIUM,
          label: strings('edit_gas_fee_eip1559.medium'),
        },
        {
          name: AppConstants.GAS_OPTIONS.HIGH,
          label: strings('edit_gas_fee_eip1559.high'),
        },
      ]
        .filter(({ name }) => !shouldIgnore(name))
        .map(({ name, label, ...option }) => ({
          name,
          label: (selectedValue: string, disabled: boolean) => (
            <Text bold primary={selectedValue && !disabled}>
              {label}
            </Text>
          ),
          topLabel: recommended?.name === name && recommended.render,
          ...option,
          ...extendOptions[name],
        })),
    [recommended, extendOptions, shouldIgnore],
  );

  const renderWarning = useMemo(() => {
    if (!warning) return null;
    if (typeof warning === 'string')
      return (
        <Alert
          small
          type={AlertType.Warning}
          renderIcon={() => (
            <MaterialCommunityIcon
              name="information"
              size={20}
              color={colors.warning.default}
            />
          )}
          style={styles.warningContainer}
        >
          {() => (
            <View style={styles.warningTextContainer}>
              <Text black style={styles.warningText}>
                {warning}
              </Text>
            </View>
          )}
        </Alert>
      );

    return warning;
  }, [warning, styles, colors]);

  const renderError = useMemo(() => {
    if (!error) return null;
    if (typeof error === 'string')
      return (
        <Alert
          small
          type={AlertType.Error}
          renderIcon={() => (
            <MaterialCommunityIcon
              name="information"
              size={20}
              color={colors.error.default}
            />
          )}
          style={styles.warningContainer}
        >
          {() => (
            <View style={styles.warningTextContainer}>
              <Text red style={styles.warningText}>
                {error}
              </Text>
            </View>
          )}
        </Alert>
      );

    return error;
  }, [error, styles, colors]);

  const {
    suggestedGasLimit,
    suggestedGasPrice,
    transactionFee,
    transactionFeeFiat,
  } = gasTransaction;

  const isMainnet = isMainnetByChainId(chainId);
  const nativeCurrencySelected = primaryCurrency === 'ETH' || !isMainnet;
  let gasFeePrimary, gasFeeSecondary;
  if (nativeCurrencySelected) {
    gasFeePrimary = transactionFee;
    gasFeeSecondary = transactionFeeFiat;
  } else {
    gasFeePrimary = transactionFeeFiat;
    gasFeeSecondary = transactionFee;
  }

  const valueToWatch = transactionFee;

  return (
    <View style={styles.root}>
      <ScrollView style={styles.wrapper}>
        <TouchableWithoutFeedback>
          <View>
            <View>
              <View style={styles.customGasHeader}>
                <TouchableOpacity onPress={onCancel}>
                  <Icon
                    name={'ios-arrow-back'}
                    size={24}
                    color={colors.text.default}
                  />
                </TouchableOpacity>
                <Text bold black>
                  {strings('transaction.edit_network_fee')}
                </Text>
                <Icon
                  name={'ios-arrow-back'}
                  size={24}
                  color={colors.background.default}
                />
              </View>
            </View>
            {renderWarning}
            {renderError}
            <FadeAnimationView
              valueToWatch={valueToWatch}
              animateOnChange={animateOnChange}
              onAnimationStart={onUpdatingValuesStart}
              onAnimationEnd={onUpdatingValuesEnd}
            >
              <View style={styles.headerContainer}>
                <View style={styles.headerTitle}>
                  <View style={styles.headerTitleSide}>
                    <Text right black style={styles.headerText}>
                      ~
                    </Text>
                  </View>
                  <Text black style={styles.headerText}>
                    {gasFeePrimary}
                  </Text>
                  <View style={styles.headerTitleSide} />
                </View>
                <Text big black>
                  <Text bold black>
                    {gasFeeSecondary}
                  </Text>
                </Text>
              </View>
              {!onlyAdvanced && (
                <View>
                  <HorizontalSelector
                    selected={selectedOption}
                    onPress={selectOption}
                    options={renderOptions}
                  />
                </View>
              )}
              <View style={styles.advancedOptionsContainer}>
                {!onlyAdvanced && (
                  <TouchableOpacity
                    onPress={toggleAdvancedOptions}
                    style={styles.advancedOptionsButton}
                  >
                    <Text noMargin link bold>
                      {strings('edit_gas_fee_eip1559.advanced_options')}
                    </Text>
                    <Text noMargin link bold style={styles.advancedOptionsIcon}>
                      <Icon
                        name={`ios-arrow-${
                          showAdvancedOptions ? 'up' : 'down'
                        }`}
                      />
                    </Text>
                  </TouchableOpacity>
                )}
                {showAdvancedOptions && (
                  <View style={styles.advancedOptionsInputsContainer}>
                    <View style={styles.rangeInputContainer}>
                      <RangeInput
                        leftLabelComponent={
                          <View style={styles.labelTextContainer}>
                            <Text black bold noMargin>
                              {strings('edit_gas_fee_eip1559.gas_limit')}{' '}
                            </Text>

                            <TouchableOpacity
                              hitSlop={styles.hitSlop}
                              onPress={() => setShowRangeInfoModal('gas_limit')}
                            >
                              <MaterialCommunityIcon
                                name="information"
                                size={14}
                                style={styles.labelInfo}
                              />
                            </TouchableOpacity>
                          </View>
                        }
                        value={suggestedGasLimit}
                        onChangeValue={changedGasLimit}
                        min={GAS_LIMIT_MIN}
                        name={strings('edit_gas_fee_eip1559.gas_limit')}
                        increment={GAS_LIMIT_INCREMENT}
                      />
                    </View>
                    <View style={styles.rangeInputContainer}>
                      <RangeInput
                        leftLabelComponent={
                          <View style={styles.labelTextContainer}>
                            <Text black bold noMargin>
                              {strings('edit_gas_fee_eip1559.gas_price')}{' '}
                            </Text>

                            <TouchableOpacity
                              hitSlop={styles.hitSlop}
                              onPress={() => setShowRangeInfoModal('gas_price')}
                            >
                              <MaterialCommunityIcon
                                name="information"
                                size={14}
                                style={styles.labelInfo}
                              />
                            </TouchableOpacity>
                          </View>
                        }
                        value={suggestedGasPrice}
                        name={strings('edit_gas_fee_eip1559.gas_price')}
                        unit={'GWEI'}
                        increment={GAS_PRICE_INCREMENT}
                        min={GAS_PRICE_MIN}
                        inputInsideLabel={
                          transactionFeeFiat && `≈ ${transactionFeeFiat}`
                        }
                        onChangeValue={changedGasPrice}
                        error={gasPriceError}
                      />
                    </View>
                  </View>
                )}
              </View>
            </FadeAnimationView>
            <View>
              <StyledButton
                type={'confirm'}
                onPress={save}
                disabled={Boolean(error) || isAnimating}
              >
                {strings('edit_gas_fee_eip1559.save')}
              </StyledButton>
            </View>
            <InfoModal
              isVisible={Boolean(showRangeInfoModal)}
              title={
                showRangeInfoModal === 'gas_limit'
                  ? strings('edit_gas_fee_eip1559.gas_limit')
                  : showRangeInfoModal === 'gas_price'
                  ? strings('edit_gas_fee_eip1559.gas_price')
                  : null
              }
              toggleModal={() => setShowRangeInfoModal(null)}
              body={
                <View>
                  <Text grey infoModal>
                    {showRangeInfoModal === 'gas_limit' &&
                      strings(
                        'edit_gas_fee_eip1559.learn_more_gas_limit_legacy',
                      )}
                    {showRangeInfoModal === 'gas_price' &&
                      strings('edit_gas_fee_eip1559.learn_more_gas_price')}
                  </Text>
                </View>
              }
            />
          </View>
        </TouchableWithoutFeedback>
      </ScrollView>
    </View>
  );
};

EditGasFeeLegacy.defaultProps = {
  ignoreOptions: [],
  warningMinimumEstimateOption: AppConstants.GAS_OPTIONS.LOW,
};

export default EditGasFeeLegacy;
