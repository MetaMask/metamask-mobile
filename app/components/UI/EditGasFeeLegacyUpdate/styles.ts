import { StyleSheet } from 'react-native';
import Device from '../../../util/device';

const createStyles = (colors: any) =>
  StyleSheet.create({
    root: {
      backgroundColor: colors.background.default,
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      minHeight: 200,
      maxHeight: '95%',
      paddingTop: 24,
      paddingBottom: Device.isIphoneX() ? 32 : 24,
    },
    wrapper: {
      paddingHorizontal: 24,
    },
    customGasHeader: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      width: '100%',
      paddingBottom: 20,
    },
    headerContainer: {
      alignItems: 'center',
      marginBottom: 22,
    },
    headerText: {
      fontSize: 48,
    },
    headerTitle: {
      flexDirection: 'row',
    },
    headerTitleSide: {
      flex: 1,
    },
    labelTextContainer: {
      flexDirection: 'row',
      alignItems: 'center',
    },
    hitSlop: {
      top: 10,
      left: 10,
      bottom: 10,
      right: 10,
    },
    labelInfo: {
      color: colors.text.muted,
    },
    advancedOptionsContainer: {
      marginTop: 25,
      marginBottom: 30,
    },
    advancedOptionsInputsContainer: {
      marginTop: 14,
    },
    rangeInputContainer: {
      marginBottom: 20,
    },
    advancedOptionsButton: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
    },
    advancedOptionsIcon: {
      paddingTop: 1,
      marginLeft: 5,
    },
    warningTextContainer: {
      paddingLeft: 4,
      lineHeight: 20,
      textAlign: 'center',
    },
    warningText: {
      lineHeight: 20,
      color: colors.text.default,
    },
  });

export default createStyles;
