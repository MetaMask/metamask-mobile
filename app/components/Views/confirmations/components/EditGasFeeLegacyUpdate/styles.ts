import { StyleSheet } from 'react-native';

import Device from '../../../../../util/device';

const createStyles = (colors: any) =>
  StyleSheet.create({
    root: {
      backgroundColor: colors.background.default,
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      minHeight: 200,
      maxHeight: '95%',
      paddingTop: 24,
      paddingBottom: Device.isIphoneX() ? 32 : 24,
    },
    wrapper: {
      paddingHorizontal: 24,
    },
    customGasHeader: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      width: '100%',
      paddingBottom: 20,
    },
    headerContainer: {
      alignItems: 'center',
      marginBottom: 22,
    },
    headerText: {
      fontSize: 48,
    },
    headerTitleSide: {
      flex: 1,
    },
    labelTextContainer: {
      flexDirection: 'row',
      alignItems: 'center',
    },
    hitSlop: {
      top: 10,
      left: 10,
      bottom: 10,
      right: 10,
    },
    labelInfo: {
      color: colors.text.muted,
    },
    advancedOptionsContainer: {
      marginTop: 25,
      marginBottom: 30,
    },
    advancedOptionsInputsContainer: {
      marginTop: 14,
    },
    rangeInputContainer: {
      marginBottom: 20,
    },
    advancedOptionsButton: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
    },
    advancedOptionsIcon: {
      paddingTop: 1,
      marginLeft: 5,
    },
    textContainer: {
      lineHeight: 20,
      textAlign: 'center',
    },
    text: {
      lineHeight: 20,
      marginHorizontal: 4,
    },
    dappEditGasContainer: {
      marginVertical: 20,
    },
  });

export default createStyles;
