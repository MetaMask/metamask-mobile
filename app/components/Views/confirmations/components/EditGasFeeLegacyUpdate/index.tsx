import BigNumber from 'bignumber.js';
/* eslint-disable react/display-name */
import React, { useCallback, useMemo, useState } from 'react';
import {
  ScrollView,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import MaterialCommunityIcon from 'react-native-vector-icons/MaterialCommunityIcons';
import { useSelector } from 'react-redux';

import { GAS_ESTIMATE_TYPES } from '@metamask/gas-fee-controller';

import { strings } from '../../../../../../locales/i18n';
import Text, {
  TextColor,
  TextVariant,
} from '../../../../../component-library/components/Texts/Text';
import { MetaMetricsEvents } from '../../../../../core/Analytics';
import { useGasTransaction } from '../../../../../core/GasPolling/GasPolling';
import { selectChainId } from '../../../../../selectors/networkController';
import {
  getDecimalChainId,
  isMainnetByChainId,
} from '../../../../../util/networks';
import { useTheme } from '../../../../../util/theme';
import Alert, { AlertType } from '../../../../Base/Alert';
import RangeInput from '../../../../Base/RangeInput';
import FadeAnimationView from '../../../../UI/FadeAnimationView';
import StyledButton from '../../../../UI/StyledButton';
import InfoModal from '../../../../UI/Swaps/components/InfoModal';
import createStyles from './styles';
import { EditGasFeeLegacyUpdateProps, EditLegacyGasTransaction } from './types';
import {
  GAS_LIMIT_INCREMENT,
  GAS_PRICE_INCREMENT,
  GAS_LIMIT_MIN,
  GAS_PRICE_MIN,
} from '../../../../../util/gasUtils';
import { useMetrics } from '../../../../../components/hooks/useMetrics';
import { selectGasFeeEstimates } from '../../../../../selectors/confirmTransaction';

const EditGasFeeLegacy = ({
  onCancel,
  onSave,
  error,
  warning,
  onUpdatingValuesStart,
  onUpdatingValuesEnd,
  animateOnChange,
  isAnimating,
  analyticsParams,
  view,
  onlyGas,
  selectedGasObject,
  hasDappSuggestedGas,
}: EditGasFeeLegacyUpdateProps) => {
  const { trackEvent } = useMetrics();
  const [showRangeInfoModal, setShowRangeInfoModal] = useState<boolean>(false);
  const [infoText, setInfoText] = useState<string>('');
  const [gasPriceError, setGasPriceError] = useState<string>('');
  const [showEditUI, setShowEditUI] = useState<boolean>(!hasDappSuggestedGas);
  const [gasObjectLegacy, updateGasObjectLegacy] = useState<{
    legacyGasLimit: string;
    suggestedGasPrice: string;
    suggestedMaxFeePerGas?: string;
  }>({
    legacyGasLimit: selectedGasObject.legacyGasLimit,
    suggestedGasPrice:
      selectedGasObject.suggestedGasPrice ||
      selectedGasObject.suggestedMaxFeePerGas,
  });

  const { colors } = useTheme();
  const styles = createStyles(colors);
  const gasFeeEstimate = useSelector(selectGasFeeEstimates);

  const primaryCurrency = useSelector(
    (state: any) => state.settings.primaryCurrency,
  );

  const gasEstimateType = useSelector(
    (state: any) =>
      state.engine.backgroundState.GasFeeController.gasEstimateType,
  );

  const chainId = useSelector((state: any) => selectChainId(state));

  const gasTransaction = useGasTransaction({
    onlyGas,
    legacy: true,
    gasObjectLegacy,
  });

  const save = useCallback(() => {
    trackEvent(MetaMetricsEvents.GAS_FEE_CHANGED, {
      ...analyticsParams,
      chain_id: getDecimalChainId(chainId),
      function_type: view,
      gas_mode: 'Basic',
    });

    const newGasPriceObject = {
      suggestedGasPrice: gasObjectLegacy?.suggestedGasPrice,
      legacyGasLimit: gasObjectLegacy?.legacyGasLimit,
    };
    onSave(gasTransaction, newGasPriceObject);
  }, [
    onSave,
    gasTransaction,
    gasObjectLegacy,
    analyticsParams,
    chainId,
    view,
    trackEvent,
  ]);

  const changeGas = useCallback((gas) => {
    updateGasObjectLegacy({
      legacyGasLimit: gas.suggestedGasLimit,
      suggestedGasPrice: gas.suggestedGasPrice,
    });
  }, []);

  const changedGasPrice = useCallback(
    (value: string) => {
      let newGas: { suggestedGasPrice: typeof value } | undefined;

      const lowerValue = new BigNumber(
        gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY
          ? gasFeeEstimate?.low
          : gasFeeEstimate?.gasPrice,
      );
      const higherValue = new BigNumber(
        gasEstimateType === GAS_ESTIMATE_TYPES.LEGACY
          ? gasFeeEstimate?.high
          : gasFeeEstimate?.gasPrice,
      ).multipliedBy(new BigNumber(1.5));

      const valueBN = new BigNumber(value);

      if (!lowerValue.isNaN() && valueBN.lt(lowerValue)) {
        setGasPriceError(strings('edit_gas_fee_eip1559.gas_price_low'));
      } else if (!higherValue.isNaN() && valueBN.gt(higherValue)) {
        setGasPriceError(strings('edit_gas_fee_eip1559.gas_price_high'));
      } else {
        setGasPriceError('');
      }

      if (typeof gasTransaction === 'object') {
        newGas = { ...gasTransaction, suggestedGasPrice: value };
      } else {
        newGas = { suggestedGasPrice: value };
      }

      changeGas(newGas);
    },
    [changeGas, gasEstimateType, gasTransaction, gasFeeEstimate],
  );

  const changedGasLimit = useCallback(
    (value: string) => {
      const newGas: { suggestedGasLimit: typeof value } =
        typeof gasTransaction === 'object'
          ? { ...gasTransaction, suggestedGasLimit: value }
          : { suggestedGasLimit: value };

      changeGas(newGas);
    },
    [changeGas, gasTransaction],
  );

  const showTransactionWarning = useMemo(() => {
    if (!warning) return null;
    if (typeof warning === 'string')
      return (
        <Alert
          small
          type={AlertType.Warning}
          renderIcon={() => (
            <MaterialCommunityIcon
              name="information"
              size={20}
              color={colors.warning.default}
            />
          )}
        >
          {() => (
            <View>
              <Text color={TextColor.Warning} style={styles.text}>
                {warning}
              </Text>
            </View>
          )}
        </Alert>
      );

    return warning;
  }, [warning, styles, colors]);

  const showTransactionError = useMemo(() => {
    if (!error) return null;
    if (typeof error === 'string')
      return (
        <Alert
          small
          type={AlertType.Error}
          renderIcon={() => (
            <MaterialCommunityIcon
              name="information"
              size={20}
              color={colors.error.default}
            />
          )}
        >
          {() => (
            <View>
              <Text color={TextColor.Error} style={styles.text}>
                {error}
              </Text>
            </View>
          )}
        </Alert>
      );

    return error;
  }, [error, styles, colors]);

  const {
    suggestedGasLimit,
    suggestedGasPrice,
    transactionFee,
    transactionFeeFiat,
  } = gasTransaction as EditLegacyGasTransaction;

  const isMainnet = isMainnetByChainId(chainId);
  const nativeCurrencySelected = primaryCurrency === 'ETH' || !isMainnet;
  let gasFeePrimary: string | null | undefined,
    gasFeeSecondary: string | null | undefined;
  if (nativeCurrencySelected) {
    gasFeePrimary = transactionFee;
    gasFeeSecondary = transactionFeeFiat;
  } else {
    gasFeePrimary = transactionFeeFiat;
    gasFeeSecondary = transactionFee;
  }

  const valueToWatch = transactionFee;

  const handleInfoModalPress = (text: string) => {
    setShowRangeInfoModal(true);
    setInfoText(text);
  };

  return (
    <View style={styles.root}>
      <ScrollView style={styles.wrapper}>
        <TouchableWithoutFeedback>
          <View>
            <View>
              <View style={styles.customGasHeader}>
                <TouchableOpacity onPress={onCancel}>
                  <Icon
                    name={'ios-arrow-back'}
                    size={24}
                    color={colors.text.default}
                  />
                </TouchableOpacity>
                <Text variant={TextVariant.HeadingSM}>
                  {strings('transaction.edit_priority')}
                </Text>
                <Icon
                  name={'ios-arrow-back'}
                  size={24}
                  color={colors.background.default}
                />
              </View>
            </View>
            {showTransactionWarning}
            {showTransactionError}

            {!showEditUI ? (
              <View style={styles.dappEditGasContainer}>
                <View style={styles.headerContainer}>
                  <Text variant={TextVariant.DisplayMD} style={{}}>
                    ~ {gasFeePrimary}
                  </Text>
                  <View style={styles.headerTitleSide} />
                  <Text
                    variant={TextVariant.BodyLGMedium}
                    color={TextColor.Alternative}
                  >
                    {gasFeeSecondary}
                  </Text>
                </View>
                <StyledButton
                  type={'orange'}
                  onPress={() => setShowEditUI(true)}
                >
                  {strings('edit_gas_fee_eip1559.edit_suggested_gas_fee')}
                </StyledButton>
              </View>
            ) : (
              <View style={styles.dappEditGasContainer}>
                <FadeAnimationView
                  valueToWatch={valueToWatch}
                  animateOnChange={animateOnChange}
                  onAnimationStart={onUpdatingValuesStart}
                  onAnimationEnd={onUpdatingValuesEnd}
                >
                  <View style={styles.headerContainer}>
                    <Text variant={TextVariant.DisplayMD} style={{}}>
                      ~ {gasFeePrimary}
                    </Text>
                    <View style={styles.headerTitleSide} />
                    <Text
                      variant={TextVariant.BodyLGMedium}
                      color={TextColor.Alternative}
                    >
                      {gasFeeSecondary}
                    </Text>
                  </View>
                  <View style={styles.rangeInputContainer}>
                    <RangeInput
                      leftLabelComponent={
                        <View style={styles.labelTextContainer}>
                          <Text variant={TextVariant.BodyMDBold}>
                            {strings('edit_gas_fee_eip1559.gas_limit')}{' '}
                          </Text>

                          <TouchableOpacity
                            hitSlop={styles.hitSlop}
                            onPress={() => handleInfoModalPress('gas_limit')}
                          >
                            <MaterialCommunityIcon
                              name="information"
                              size={14}
                              style={styles.labelInfo}
                            />
                          </TouchableOpacity>
                        </View>
                      }
                      value={suggestedGasLimit}
                      onChangeValue={changedGasLimit}
                      min={GAS_LIMIT_MIN}
                      name={strings('edit_gas_fee_eip1559.gas_limit')}
                      increment={GAS_LIMIT_INCREMENT}
                    />
                  </View>
                  <View style={styles.rangeInputContainer}>
                    <RangeInput
                      leftLabelComponent={
                        <View style={styles.labelTextContainer}>
                          <Text variant={TextVariant.BodyMDBold}>
                            {strings('edit_gas_fee_eip1559.gas_price')}{' '}
                          </Text>
                          <Text color={TextColor.Alternative}>(GWEI) </Text>

                          <TouchableOpacity
                            hitSlop={styles.hitSlop}
                            onPress={() => handleInfoModalPress('gas_price')}
                          >
                            <MaterialCommunityIcon
                              name="information"
                              size={14}
                              style={styles.labelInfo}
                            />
                          </TouchableOpacity>
                        </View>
                      }
                      value={suggestedGasPrice}
                      name={strings('edit_gas_fee_eip1559.gas_price')}
                      increment={GAS_PRICE_INCREMENT}
                      min={GAS_PRICE_MIN}
                      inputInsideLabel={
                        transactionFeeFiat && `≈ ${transactionFeeFiat}`
                      }
                      onChangeValue={changedGasPrice}
                      error={gasPriceError}
                    />
                  </View>
                </FadeAnimationView>
                <View>
                  <StyledButton
                    type={'confirm'}
                    onPress={save}
                    disabled={Boolean(error) || isAnimating}
                  >
                    {strings('edit_gas_fee_eip1559.save')}
                  </StyledButton>
                </View>
              </View>
            )}
            <InfoModal
              isVisible={Boolean(showRangeInfoModal)}
              title={
                infoText === 'gas_limit'
                  ? strings('edit_gas_fee_eip1559.gas_limit')
                  : infoText === 'gas_price'
                  ? strings('edit_gas_fee_eip1559.gas_price')
                  : null
              }
              toggleModal={() => setShowRangeInfoModal(false)}
              body={
                <View>
                  <Text color={TextColor.Alternative}>
                    {infoText === 'gas_limit' &&
                      strings(
                        'edit_gas_fee_eip1559.learn_more_gas_limit_legacy',
                      )}
                    {infoText === 'gas_price' &&
                      strings('edit_gas_fee_eip1559.learn_more_gas_price')}
                  </Text>
                </View>
              }
            />
          </View>
        </TouchableWithoutFeedback>
      </ScrollView>
    </View>
  );
};

export default EditGasFeeLegacy;
