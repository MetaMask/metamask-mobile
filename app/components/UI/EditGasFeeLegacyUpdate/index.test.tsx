import React from 'react';
import { shallow } from 'enzyme';
import EditGasFeeLegacyUpdate from './';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

const mockStore = configureMockStore();
const initialState = {
  engine: {
    backgroundState: {
      CurrencyRateController: {
        currentCurrency: 'usd',
        conversionRate: 0.1,
      },
      NetworkController: {
        provider: {
          ticker: 'ETH',
          chainId: '1',
        },
      },
    },
  },
};
const store = mockStore(initialState);

describe('EditGasFeeLegacyUpdate', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <EditGasFeeLegacyUpdate view={'Test'} />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
