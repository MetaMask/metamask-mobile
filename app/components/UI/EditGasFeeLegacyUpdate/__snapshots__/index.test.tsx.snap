// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGasFeeLegacyUpdate should render correctly 1`] = `
<EditGasFeeLegacy
  ignoreOptions={Array []}
  view="Test"
  warningMinimumEstimateOption="low"
/>
`;
